lib/Dialect/Seq/Transforms/RegOfVecToMem.cpp

```diff
@@ -0,0 +1,245 @@
+public:
+  void runOnOperation() override;
+
+private:
+  bool analyzeMemoryPattern(FirRegOp reg, MemoryPattern &pattern);
+  bool createFirMemory(MemoryPattern &pattern);
+  bool isArrayType(Type type);
+  std::optional<std::pair<uint64_t, uint64_t>> getArrayDimensions(Type type);
+
+  SmallVector<Operation *> opsToErase;
+};
+
+} // end anonymous namespace
+
+bool RegOfVecToMemPass::isArrayType(Type type) {
+  return isa<hw::ArrayType, hw::UnpackedArrayType>(type);
+}
+
+std::optional<std::pair<uint64_t, uint64_t>>
+RegOfVecToMemPass::getArrayDimensions(Type type) {
+  if (auto arrayType = dyn_cast<hw::ArrayType>(type)) {
+    auto elemType = arrayType.getElementType();
+    if (auto intType = dyn_cast<IntegerType>(elemType)) {
+      return std::make_pair(arrayType.getNumElements(), intType.getWidth());
+    }
+  }
+  return std::nullopt;
+}
+
+bool RegOfVecToMemPass::analyzeMemoryPattern(FirRegOp reg,
+                                             MemoryPattern &pattern) {
+  LLVM_DEBUG(llvm::dbgs() << "Analyzing register: " << reg << "\n");
+
+  // Check if register has array type
+  if (!isArrayType(reg.getType()))
+    return false;
+
+  pattern.memReg = reg;
+  pattern.clock = reg.getClk();
+
+  // Find the mux that drives this register
+  auto nextValue = reg.getNext();
+  auto mux = nextValue.getDefiningOp<comb::MuxOp>();
+  if (!mux)
+    return false;
+
+  LLVM_DEBUG(llvm::dbgs() << "  Found driving mux: " << mux << "\n");
+  pattern.writeMux = mux;
+
+  // Analyze mux inputs: sel ? write_result : current_memory
+  Value writeResult = mux.getTrueValue();
+  Value currentMemory = mux.getFalseValue();
+
+  // Check if false value is the current register (feedback)
+  if (currentMemory != reg.getResult())
+    return false;
+
+  // Look for array_inject operation in write path
+  auto arrayInject = writeResult.getDefiningOp<hw::ArrayInjectOp>();
+  if (!arrayInject)
+    return false;
+
+  LLVM_DEBUG(llvm::dbgs() << "  Found array_inject: " << arrayInject << "\n");
+  pattern.writeAccess = arrayInject;
+  pattern.writeAddr = arrayInject.getIndex();
+  pattern.writeData = arrayInject.getElement();
+  pattern.writeEnable = mux.getCond();
+
+  // Look for read pattern - find array_get users
+  for (auto *user : reg.getResult().getUsers()) {
+    if (auto arrayGet = dyn_cast<hw::ArrayGetOp>(user)) {
+      LLVM_DEBUG(llvm::dbgs() << "  Found array_get: " << arrayGet << "\n");
+      pattern.readAccess = arrayGet;
+      pattern.readAddr = arrayGet.getIndex();
+
+      // Check if read goes through output register
+      for (auto *readUser : arrayGet.getResult().getUsers()) {
+        if (auto outputReg = dyn_cast<FirRegOp>(readUser)) {
+          if (outputReg.getClk() == pattern.clock) {
+            LLVM_DEBUG(llvm::dbgs()
+                       << "  Found output register: " << outputReg << "\n");
+            pattern.outputReg = outputReg;
+            break;
+          }
+        }
+      }
+      break;
+    }
+  }
```

Comment:
This loop might have to go through every user to ensure that there's only a single `array_get` using the register. Otherwise, replacing the register would potentially break because there are other users.

It's probably worthwhile to add a test that has a `seq.firreg` that looks like it qualifies for `seq.firmem` replacement, but has its value returned as an output or some other use besides an `hw.array_get`, which should then block the replacement.